core/modules.rs

```diff
@@ -55,209 +49,269 @@ pub trait Loader: Send + Sync {
-          // 1. Source code resolved result have the same module name as requested
-          //    and is not yet registered
-          //     -> register
-          // 2. Source code resolved result have a different name as requested:
-          //   2a. The module with resolved module name has been registered
-          //     -> alias
-          //   2b. The module with resolved module name has not yet been registerd
-          //     -> register & alias
-          let is_module_registered = {
-            let modules = self.modules.lock().unwrap();
-            modules.is_registered(&source_code_info.module_name)
-          };
-
-          let need_alias = source_code_info.module_name != completed.url;
-
-          if !is_module_registered {
-            let module_name = &source_code_info.module_name;
-
-            let mod_id = {
-              let isolate = self.isolate.lock().unwrap();
-              isolate.mod_new(
-                completed.is_root,
-                module_name,
-                &source_code_info.code,
-              )
-            }?;
-
-            if completed.is_root {
-              assert!(self.root_id.is_none());
-              self.root_id = Some(mod_id);
-            }
 
-            // Register new module.
-            {
-              let mut modules = self.modules.lock().unwrap();
-              modules.register(mod_id, module_name);
-              // If necessary, register the alias.
-              if need_alias {
-                let module_alias = &completed.url;
-                modules.alias(module_alias, module_name);
-              }
-            }
+    // If we just finished loading the root module, store the root module id.
+    match self.state {
+      State::LoadingRoot => self.state = State::LoadingImports(module_id),
+      State::LoadingImports(..) => {}
+      _ => unreachable!(),
+    };
 
-            // Now we must iterate over all imports of the module and load them.
-            let imports = {
-              let isolate = self.isolate.lock().unwrap();
-              isolate.mod_get_imports(mod_id)
-            };
-            let referrer = module_name;
-            for specifier in imports {
-              self.add(&specifier, referrer, Some(mod_id))?;
-            }
-          } else if need_alias {
-            let mut modules = self.modules.lock().unwrap();
-            modules.alias(&completed.url, &source_code_info.module_name);
+    // If all imports have been loaded, instantiate the root module.
+    if self.pending.is_empty() {
+      let root_id = match self.state {
+        State::LoadingImports(mod_id) => mod_id,
+        _ => unreachable!(),
+      };
+
+      let mut resolve_cb =
+        |specifier: &str, referrer_id: deno_mod| -> deno_mod {
+          let modules = self.modules.lock().unwrap();
+          let referrer = modules.get_name(referrer_id).unwrap();
+          match self.loader.resolve(specifier, &referrer, false) {
```

Comment:
Question: is `resolve_cb` called only for root module? I believe during writing import maps I asserted it is root module, but I'm not sure now. If it's not a root then hardcoded `false` should be changed

Reply:
correct. (but now the root module is no longer the same as the main module - it's just the root of any RecursiveLoad)

Reply:
> (but now the root module is no longer the same as the main module - it's just the root of any RecursiveLoad)

So if I understand correctly then we should change that `false` to actual comparison if this module is main module

Reply:
Done. We'll need to have some tests soon showing import maps working against dynamic import - this change had no effect on the tests.